### ISSUE TITLE
Avoid MPI_Init for Serial I/O when built with MPI

### ISSUE THREAD

**Issue by ax3l** [ornladios/ADIOS2#1443]
Hi,

we discussed this offline before and we are currently at the point where we would need the feature again. When compiling ADIOS2 with MPI features, it would be wonderful if the serial constructors would still be exposed.

https://github.com/ornladios/ADIOS2/blob/02f246eb328a524eefd00286ce88ed1c0f2172f9/bindings/CXX11/cxx11/ADIOS.h#L67-L80

This would allow users to use a cluster module with MPI support and build small serial tools against it, e.g. for meta-data reads, without the need to call those with `mpirun` (and without `MPI_Init`, ideally).

Is it possible to expose the MPI functionality as a true superset of the serial functionality? Also from a package management point of view, this would make a lot of sense for compatibility.

Regarding the current implementation in [`source/adios2/ADIOSMPI.h`](https://github.com/ornladios/ADIOS2/blob/7b3939beec73029f0fe7e62e54f29e0832e697fa/source/adios2/ADIOSMPI.h), I am a bit afraid we run into usability issues (also seen in practical examples with ADIOS1: https://github.com/ornladios/ADIOS/issues/183). Instead of mocking the MPI APIs, could we potentially just use those as a wrapper and an internal state (MPI-parallel or serial) to switch the implementation at runtime?

**Comment by williamfgc** [comment]
@ax3l a couple of things:

1) in ADIOS2 the way we currently write portable code is using the macro `#ifdef ADIOS2_HAVE_MPI` 

```
#ifdef ADIOS2_HAVE_MPI
    adios2::ADIOS adios(comm);
#else 
    adios2::ADIOS adios;
#endif
```

2) source/adios2/ADIOSMPI.h is completely private, never exposed to the application. In ADIOS1.x mpidummy.h was exposed while in ADIOS2 is private and inside the `namespace adios2`

I don't know how a truly non-conflicting single MPI and Non-MPI library can be deployed. Ideas are welcome or, even better, we can borrow from an existing implementation doing that. Thanks!

**Comment by williamfgc** [comment]
@ax3l a couple of things:

1) in ADIOS2 the way we currently write portable code is using the macro `#ifdef ADIOS2_HAVE_MPI` 

```
#ifdef ADIOS2_HAVE_MPI
    adios2::ADIOS adios(comm);
#else 
    adios2::ADIOS adios;
#endif
```

2) source/adios2/ADIOSMPI.h is completely private, never exposed to the application. In ADIOS1.x mpidummy.h was exposed while in ADIOS2 is private and inside the `namespace adios2`

I don't know how a truly non-conflicting single MPI and Non-MPI library can be deployed. Ideas are welcome or, even better, we can borrow from an existing implementation doing that. Thanks!

**Comment by ax3l** [comment]
The idea I had was the following: maybe we can put a state in the core implementation classes such as `core::ADIOS`, `Engine::Engine`, `Transport::Transport`, etc. that stores if `MPI_Initialized` is `false` and the communicator `MPI_COMM_SELF` was passed. Then, instead of calling `MPI_Comm_rank/size` and the other MPI functions seen in `source/adios2/ADIOSMPI.h` directly we can add a small wrapper that selects the MPI implementation from the dummy or the real ones based on this state (at runtime rather than compile time)? That would allow to avoid calling `MPI_init`, e.g. in `bpls`.

**Comment by williamfgc** [comment]
@ax3l  at the public interface of `adios2::ADIOS` the `MPI_COMM_SELF` definition must come from the user. I understand what you mean for the above internal objects. Not sure if you can pass `MPI_COMM_SELF` or any MPI defintion without inclusion of `mpi.h` unless a new definition is passed which might conflict with actual MPI definitions if somewhere in the pipeline at linking. That was the adios1.x issue when mixing MPI code with the non-MPI library, thus we decided to completely separate interfaces at compile time and keep mpidummy.h inside `namespace adios2`. It would be helpful if we can have a simple example as I don't have a clear understanding how this would work at the public interface level. Thanks!

**Comment by williamfgc** [comment]
The way I have worked with MPI libraries that allow serial versions is to play with their MACRO `XXX_USE_MPI` or `XXX_HAVE_MPI`  I am currently doing that with MFEM and VTK. I am personally open to more convenient ways.

**Comment by williamfgc** [comment]
to narrow it down:

```
#include <adios2.h>

int main()
{
// here MPI_Comm in ADIOS.h or ADIOS.cpp (if forward declared) must come from somewhere at the linking stage....without MPI_Init (or mpi.h) any non-MPI definition of MPI_Comm will conflict if MPI is linked
adios2::ADIOS adios;
}
```

I am not aware of a MPI/serial library that can by-pass the above conflict using a single library. Unless two separate libraries and appropriate headers are provided. This is the piece I am don't know enough about.

**Comment by ax3l** [comment]
Yes, the absolute cleanest way would be a second header, say `<adios2-serial.h>` and a separate library with no MPI-related libs.

For practial cases, it's still okay if in the `USE_MPI` case the `<mpi.h>` header is still transitively pulled and the MPI lib still needs to be linked, e.g. since it's available anyway from a HPC system's module and will be covered by `adios-config`/`ADIOS2Config.cmake` without the user noticing much.

The "serial" functionality, [e.g. the constructor](https://github.com/ornladios/ADIOS2/blob/7b3939beec73029f0fe7e62e54f29e0832e697fa/source/utils/bpls/bpls.cpp#L1187), should be still available and can compile against the usual MPI library that ADIOS2 was build with. Yet internally, it switches to the `adios::helper::mpi` functionality if `MPI_COMM_SELF` is passed (from the internal constructor) *and* `MPI_Finalize` is found to be `false` as well.

Basically, just like `bpsl` is written but without the `#ifdef ADIOS2_HAVE_MPI` branches.

This will enable several nice cases:
- serial tools can be build with the same module combination on a HPC system as a main application
- `mpirun` is avoided where this is allowed and possible
- Debian/Suse/... binary packages such as `adios2-serial adios2-openmpi adios2-mpich` are not fully incompatible as providers for serial tools, instead, a serial tool can depend on any of those
- when linking a dependency that pulls serial functionality of ADIOS2 with a dependency that pulls mpi-enabled functionality of ADIOS2, those ADIOS2 libs and symbols are not incompatible and symbol hiding in wrapper libraries can be avoided for end users

**Comment by williamfgc** [comment]
@ax3l I see the benefits, but the issue I am bringing up is conflicting definitions at the public layer.
Let me clarify, the ADIOS class has a member MPI_Comm which must be defined when linking (when adios2 was compiled with mpi, it came from `mpi.h`):

```
class  ADIOS
{
public:
MPI_Comm m_Comm;

}
```
without `mpi.h` users will need to provide a conflicting definition which work work. Can `MPI_Comm` be defined without a `MPI_Init`? That's the piece that I am missing.

**Comment by ax3l** [comment]
Yes, I understand what you mean. I think that would be harder to do since `MPI_Comm` and its memory layout is not defined in the MPI standard and thus the `ADIOS` public ABI of the class cannot be fully defined. Maybe passing a not typesafe `void*` would be an option if one wants to avoid two `ADIOS` classes and two headers.

**Comment by germasch** [comment]
To put in my 2 cents, while I agree that it'd be nice to have adios2 work with and without MPI simultaneously, I think it's technically rather cumbersome to do.

It's not a header-only library, so one can't just switch MPI implementations easily (e.g., sizeof(MPI_Comm) isn't even constant across different MPI implementations). So essentially, the compiled code has to be different -- and hence an `ADIOS` object created from the serial constructor will be different from an `ADIOS` object from a constructor that takes an MPI communicator. That's not really possible, unless you make them different classes by putting them into different namespaces.

One way I see that things could be done would be to compile a big library that essentially combines a serial and a parallel library into one big .so, using different namespaces (or different version trickery or something in the .so). And then one could have a header that depending on USE_MPI does something like `using ADIOS = serial::ADIOS` or `using ADIOS = parallel::ADIOS`. That seems quite hacky to me, and not easy to implement, and that library will still pull in the MPI library, anyway, if though it may not use it.

One solution I can think of is an "install both" mode, which would compile both libadios2.so and libadios2_serial.so, and provides `adios2/adios2.h" and "adios2_serial.h", and one can then include one or the other and correspondingly link with one or the other. But I don't really like this, either...

Thinking about this a bit more, though, I think HDF5 if compiled parallel let's you use it serially without trouble. So maybe there's a point in considering that. But it won't be straightforward to achieve this in adios2, I don't think. The idea would be to only use an MPI_Comm when it's actually required, not for the top level objects.

**Comment by ax3l** [comment]
Actually the initial point of this issue is just about your last paragraph for exactly these reasons. Keep the interface as is, including MPI headers if they were present, but always expose the "serial" constructor as well and avoid `MPI_` calls if MPI is found uninitialized, so we do not need to artificially initialize MPI.

**Comment by williamfgc** [comment]
@germasch thanks, yes that's pretty much how I see it. Would be helpful to an example to how HDF5 does it. My understanding is that when adios2 is compiled with MPI, the latter becomes a public dependency (any MPI-based library for that matter). @ax3l at this point, a proof-of-concept that passes the linking stage would be helpful.

**Comment by germasch** [comment]
I can implement it. A simple proof-of-concept is very simple:

```diff
diff --git a/bindings/CXX11/cxx11/ADIOS.cpp b/bindings/CXX11/cxx11/ADIOS.cpp
index 1d761440..0e93d76c 100644
--- a/bindings/CXX11/cxx11/ADIOS.cpp
+++ b/bindings/CXX11/cxx11/ADIOS.cpp
@@ -25,6 +25,8 @@ ADIOS::ADIOS(MPI_Comm comm, const bool debugMode) : ADIOS("", comm, debugMode)
 {
 }

+ADIOS::ADIOS(const bool debugMode) : ADIOS("", MPI_COMM_SELF, debugMode) {}
+
 #else
 ADIOS::ADIOS(const std::string &configFile, const bool debugMode)
 : m_ADIOS(std::make_shared<core::ADIOS>(configFile, debugMode, "C++"))
diff --git a/bindings/CXX11/cxx11/ADIOS.h b/bindings/CXX11/cxx11/ADIOS.h
index dff4a599..7f3a8372 100644
--- a/bindings/CXX11/cxx11/ADIOS.h
+++ b/bindings/CXX11/cxx11/ADIOS.h
@@ -79,6 +79,8 @@ public:
      */
     ADIOS(const std::string &configFile, const bool debugMode = true);

+#endif
+
     /**
      * Starting point for non-MPI apps. Creates an ADIOS object
      * @param debugMode true: extra user-input debugging information, false: run
@@ -87,8 +89,7 @@ public:
      * incorrect
      */
     ADIOS(const bool debugMode = true);
-#endif
-
+
     /** object inspection true: valid object, false: invalid object */
     explicit operator bool() const noexcept;
```

This will enable an application to use the serial constructor even when ADIOS2 is built with MPI. ADIOS2 will, of course, still pull in MPI, which is not much of a problem when it's provided as a shared library, because that'll happen automatically.

However, it won't really work, because once the app is actually using such an ADIOS object, MPI functions will be called, and they'll complain because MPI has not been initialized.

So to make this actually work, I see two options:
* have ADIOS2 itself call MPI_Init in such a case (I think that's just too hacky, though).
* in ADIOS2, wrap a bunch of MPI functions always, ie, MPI_Comm_dup, MPI_Comm_rank/size, MPI_Bcast, maybe more, so that they'll not call down into MPI when it's not initialized. 

(The third option would be a more thorough reengineering of ADIOS2 to move down MPI calls down to only the layers where they're actually needed, which I think is too big of a change to be realistic).

**Comment by williamfgc** [comment]
@germasch thanks! This basically confirms that it's not feasible that a single library (compiled with MPI) can support such dual behavior. MPI-based libraries wouldn't even link if MPI itself is not linked, unless we introduce conflicting MPI symbols (a hack again) which will lead to @ax3l issue list with mpidummy.h in adios1. 

BTW, ADIOS2 is already doing the third option. MPI calls are always done privately. The issue is narrowed down to linking the public API handshake with MPI (passing a comm) that any MPI-based library requires. The cleanest solution would be to keep the serial and parallel implementations completely separately, which most projects I am aware of keep track at compile time via a MACRO, as it's not feasible to do at run time once a library is compiled with MPI. To make things worse MPI distributions are further split between openmpi and mpich based distributions that are incompatible. 

@ax3l while I share the same wishlist, I believe this is not feasible with a single library. If you know a better (working) way please let us know.

**Comment by germasch** [comment]
I don't think you understood me right. What the patch shows is the providing that API is easily doable. But that simple change doesn't actually make it usable without calling MPI_Init. What I said, though, is that it's possible to make that work quite easily, but it will require some (relatively minimal) wrapping of MPI inside of adios2.

**Comment by williamfgc** [comment]
Apologies if I wasn't clear enough, by feasible I mean working: compiled/linked/running/tested, not only the code interface.

**Comment by ax3l** [comment]
My primary goal would be to get `bpls` running in the sense that it does not require the `MPI_Init`/`MPI_Finalize` anymore if ADIOS2 is build with MPI.

Here is a quick hack that starts this, but we should do this more systematic in all places:
```diff
diff --git a/source/adios2/core/IO.cpp b/source/adios2/core/IO.cpp
index 238c8c93..e6a939a4 100644
--- a/source/adios2/core/IO.cpp
+++ b/source/adios2/core/IO.cpp
@@ -438,7 +438,18 @@ Engine &IO::Open(const std::string &name, const Mode mode,
     }
 
     MPI_Comm mpiComm;
-    MPI_Comm_dup(mpiComm_orig, &mpiComm);
+    int flag;
+    MPI_Initialized(&flag);
+    if (flag)
+    {
+        MPI_Comm_dup(mpiComm_orig, &mpiComm);
+        // m_NeedMPICommFree = true;
+    }
+    else
+    {
+        mpiComm = mpiComm_orig;
+        // m_NeedMPICommFree = false;
+    }
     std::shared_ptr<Engine> engine;
     const bool isDefaultEngine = m_EngineType.empty() ? true : false;
     std::string engineTypeLC = m_EngineType;
diff --git a/source/adios2/toolkit/format/bp3/BP3Base.cpp b/source/adios2/toolkit/format/bp3/BP3Base.cpp
index 3f484cb4..cf419d48 100644
--- a/source/adios2/toolkit/format/bp3/BP3Base.cpp
+++ b/source/adios2/toolkit/format/bp3/BP3Base.cpp
@@ -34,7 +34,7 @@ const std::map<int, std::string> BP3Base::m_TransformTypesToNames = {
     {transform_zfp, "zfp"},           {transform_mgard, "mgard"},
     // {transform_zlib, "zlib"},
     //    {transform_bzip2, "bzip2"},
-    //    {transform_szip, "szip"},
+    //    {transform_szip, "szip"}
     //    {transform_isobar, "isobar"},
     //    {transform_aplod, "aplod"},
     //    {transform_alacrity, "alacrity"},
@@ -47,8 +47,14 @@ const std::map<int, std::string> BP3Base::m_TransformTypesToNames = {
 BP3Base::BP3Base(MPI_Comm mpiComm, const bool debugMode)
 : m_MPIComm(mpiComm), m_DebugMode(debugMode)
 {
-    MPI_Comm_rank(m_MPIComm, &m_RankMPI);
-    MPI_Comm_size(m_MPIComm, &m_SizeMPI);
+    int flag;
+    MPI_Initialized(&flag);
+    m_RankMPI = 0;
+    m_SizeMPI = 1;
+    if (flag) {
+        MPI_Comm_rank(m_MPIComm, &m_RankMPI);
+        MPI_Comm_size(m_MPIComm, &m_SizeMPI);
+    }
     m_Profiler.IsActive = true; // default
 }
 
diff --git a/source/adios2/toolkit/transport/Transport.cpp b/source/adios2/toolkit/transport/Transport.cpp
index 83e8e8ca..ce8b5f4b 100644
--- a/source/adios2/toolkit/transport/Transport.cpp
+++ b/source/adios2/toolkit/transport/Transport.cpp
@@ -20,8 +20,14 @@ Transport::Transport(const std::string type, const std::string library,
                      MPI_Comm mpiComm, const bool debugMode)
 : m_Type(type), m_Library(library), m_MPIComm(mpiComm), m_DebugMode(debugMode)
 {
-    MPI_Comm_rank(m_MPIComm, &m_RankMPI);
-    MPI_Comm_size(m_MPIComm, &m_SizeMPI);
+    int flag;
+    MPI_Initialized(&flag);
+    m_RankMPI = 0;
+    m_SizeMPI = 1;
+    if (flag) {
+        MPI_Comm_rank(m_MPIComm, &m_RankMPI);
+        MPI_Comm_size(m_MPIComm, &m_SizeMPI);
+    }
 }
 
 void Transport::IWrite(const char *buffer, size_t size, Status &status,
diff --git a/source/utils/bpls/bpls.cpp b/source/utils/bpls/bpls.cpp
index c2054cdc..1914fcd0 100644
--- a/source/utils/bpls/bpls.cpp
+++ b/source/utils/bpls/bpls.cpp
@@ -2934,7 +2934,7 @@ char *mystrndup(const char *s, size_t n)
 int main(int argc, char *argv[])
 {
 #ifdef ADIOS2_HAVE_MPI
-    MPI_Init(&argc, &argv);
+//    MPI_Init(&argc, &argv);
 #endif
     int retval = 1;
     try
@@ -2947,7 +2947,7 @@ int main(int argc, char *argv[])
         std::cout << e.what() << std::endl;
     }
 #ifdef ADIOS2_HAVE_MPI
-    MPI_Finalize();
+//    MPI_Finalize();
 #endif
     return retval;
 }
(END)
+    if (flag) {
+        MPI_Comm_rank(m_MPIComm, &m_RankMPI);
+        MPI_Comm_size(m_MPIComm, &m_SizeMPI);
+    }
 }
 
 void Transport::IWrite(const char *buffer, size_t size, Status &status,
diff --git a/source/utils/bpls/bpls.cpp b/source/utils/bpls/bpls.cpp
index c2054cdc..1914fcd0 100644
--- a/source/utils/bpls/bpls.cpp
+++ b/source/utils/bpls/bpls.cpp
@@ -2934,7 +2934,7 @@ char *mystrndup(const char *s, size_t n)
 int main(int argc, char *argv[])
 {
 #ifdef ADIOS2_HAVE_MPI
-    MPI_Init(&argc, &argv);
+//    MPI_Init(&argc, &argv);
 #endif
     int retval = 1;
     try
@@ -2947,7 +2947,7 @@ int main(int argc, char *argv[])
         std::cout << e.what() << std::endl;
     }
 #ifdef ADIOS2_HAVE_MPI
-    MPI_Finalize();
+//    MPI_Finalize();
 #endif
     return retval;
 }
```

We can avoid the recurring `MPI_Initialized(&flag);` check with a small wrapper around the MPI libs that are already identified in `mpidummy.h`

**Comment by williamfgc** [comment]
@ax3l I understand how this is done privately...my concern is very simple: would it link with serial code? the MPI_Comm symbols gets resolved from MPI when adios2 is compiled, but MPI will be missed when linking to serial code. Seems I am not getting the message somehow and showing code only won't guarantee this is possible at all. If you're correct and `bpls` can in fact run, I am all ears.

**Comment by ax3l** [comment]
I understand that and it's still needed to link against MPI although it will then not be used in communication APIs. My primary concern is really just to avoid `MPI_Init`, it's fine if a user still has to link an unused MPI, because that will allow to use the same (parallel) adios module/package for both serial python hackery, middleware and parallel tools. The additionally (unused but needed) MPI linkage is passed by `adios-config`/`ADIOS2Config.cmake` anyway, so this does not change the current state.

**Comment by germasch** [comment]
FWIW, `MPI_Comm` is not a symbol, it'd never be unresolved now matter what you link to or not. But yes, there would be actual unresolved symbols, like`MPI_Comm_dup` if not linking with MPI, even if those functions would never actually end up being called at run time. But that's an issue for building an executable, which is already resolved via the current mechanisms (e.g., cmake / adios2-config). @ax3l's patch is exactly what I mean, except, as he says, this should be done via wrappers rather than throughout the code.

**Comment by pnorbert** [comment]
After some discussion with @wfgodoy, we propose to solve what @ax3l
<https://github.com/ax3l> is asking for: write a complete abstract class
for communication and have to implementation of it, one with real MPI and
one with dummy MPI. The appropriate communicator object will be created in
the ADIOS constructor depending on if the ADIOS constructor is called with
a (real) MPI comm or without.

This will ensure that
- we have one library with both parallel and serial running capability.
- apps must be linked with MPI even though they never use it.
- serial apps linked with this library will run on login nodes that have no
MPI support

It removes the possibility to call ADIOS without any comm and Open with a
comm, but that was never a pretty option anyway. We are glad it will be
gone.

How do you feel about this?

**Comment by ax3l** [comment]
Sounds good to me, great!

I do not fully understand the last paragraph, since we can still keep a serial version without a communicator in the public interface, imho.

**Comment by williamfgc** [comment]
I think we have to narrow down some concepts we are mixing after talking to @pnorbert :

1) Serial:  @axl by serial you mean code linked against adios2 and mpi, but running without MPI_Init() , I think we agree that truly serial code (gcc/g++  without -libmpi and, nevertheless mpicc/mpic++) would never work with a library compiled against MPI publicly.

2) Cray login nodes: would 1) still run on login nodes since the don't accept MPI binaries? (this is a question, I told @pnorbert I'd test on Titan)

3) @germasch `MPI_Comm` would be replaced by the corresponding opaque pointer implementation alias (`ompi_communicator_t*`). My point is MPI needs to be publicly linked against the binary due to its public nature in adios2 to follow 1).

**Comment by ax3l** [comment]
1. yes and yes :)

2. Cray is tricky, if the lib is compiled with a generic arch this might work. 1) might actually address this, since `MPI_Init` is the call that fails, afair: https://github.com/PDC-support/mpi-lab-exercises

**Comment by williamfgc** [comment]
@ax3l great! Thanks for providing this, that addresses one my concerns! Yes, then what @pnorbert proposed should be feasible for `bpls` and Python code on Titan.  We should find a better way to name the `mpi-linked without MPI_Init()` special executable/script. Thanks!

**Comment by germasch** [comment]
See PR #1466 for a proposed solution.

**Comment by chuckatkins** [comment]
@williamfgc @pnorbert can we close this now as fixed?

**Comment by germasch** [comment]
@chuckatkins I'd say it's mostly fixed with some caveats, so I'd like to see #1480 merged (or at least discussed), too.

**Comment by germasch** [comment]
As #1480 has been merged, as far as i'm concerned, this issue can be closed, @chuckatkins .

**Comment by ax3l** [comment]
Looks good to me as well, really happy this works! Thanks a lot as well to @germasch for much of the implementation! :)
